ArduSub/control_motordetect.cpp

```diff
@@ -0,0 +1,175 @@
+ *
+ *  This mode pulses all thrusters to detect if they need to be reversed.
+ *  This still requires that the user has the correct frame selected and the motors
+ *  are connected to the correct ESCs.
+ *
+ *  For each motor:
+ *      wait until vehicle is stopped for > 500ms
+ *      apply throttle up for 500ms
+ *      If results are good:
+ *          save direction and try the next motor.
+ *      else
+ *          wait until vehicle is stopped for > 500ms
+ *          apply throttle down for 500ms
+ *          If results are good
+ *              save direction and try the next motor.
+ *          If results are bad
+ *              Abort!
+ */
+
+// controller states
+enum test_state {
+    STANDBY,
+    SETTLING,
+    THRUSTING,
+    DETECTING,
+    DONE
+};
+
+enum direction {
+    UP = 1,
+    DOWN = -1
+};
+
+static uint32_t settling_timer;
+static uint32_t thrusting_timer;
+static uint8_t md_state;
+static uint8_t current_motor;
+static int16_t current_direction;
+
+
+bool Sub::motordetect_init()
+{
+    current_motor = 0;
+    md_state = STANDBY;
+    current_direction = UP;
+    return true;
+}
+
+void Sub::motordetect_run()
+{
+    // if not armed set throttle to zero and exit immediately
+    if (!motors.armed()) {
+        motors.set_desired_spool_state(AP_Motors::DesiredSpoolState::GROUND_IDLE);
+        // Force all motors to stop
+        for(uint8_t i=0; i < AP_MOTORS_MAX_NUM_MOTORS; i++) {
+            if (motors.motor_is_enabled(i)) {
+                motors.output_test_num(i, 1500);
+            }
+        }
+        md_state = DONE;
+        return;
+    }
+
+    switch(md_state) {
+    // Motor detection is not running, set it up to start.
+    case STANDBY:
+        current_direction = UP;
+        current_motor = 0;
+        settling_timer = AP_HAL::millis();
+        md_state = SETTLING;
+        break;
+
+    // Wait until sub stays for 500ms not spinning and leveled.
+    case SETTLING:
+        // Force all motors to stop
+        for (uint8_t i=0; i <AP_MOTORS_MAX_NUM_MOTORS; i++) {
+            if (motors.motor_is_enabled(i)) {
+                !motors.output_test_num(i, 1500);
+            }
+        }
+        // wait until roll and pitch are between -10 and 10 degrees
+        // and gyro product is under a certain(experimental) threshold
+        if (abs(ahrs.roll_sensor) > 1000
+                || abs(ahrs.roll_sensor) > 1000
+                || (ahrs.get_gyro()*ahrs.get_gyro()) > 0.01) {
+            settling_timer = AP_HAL::millis();
+        }
+        // then wait 500ms more
+        if (AP_HAL::millis() > (settling_timer + 500)) {
+            md_state = THRUSTING;
+            thrusting_timer = AP_HAL::millis();
+        }
+
+        break;
+
+    // Thrusts motor for 500ms
+    case THRUSTING:
+        if (AP_HAL::millis() < (thrusting_timer + 500)) {
+            if (!motors.output_test_num(current_motor, 1500 + 300*current_direction)) {
```

Comment:
We should be detecting and thrusting simultaneously. We should maybe ramp thrust so that we just use the minimum thrust to get a signal.

Reply:
I don't like the idea of ramping because that may cause a longer movement and higher change of hitting things.
detecting and thrusting simultaneously sounds good, I'll try that.